Edit menu allows ZAP mode change this closes zaproxy/zaproxy#2375

rajiv65 opened the pull request:
- Edit menu now has an option to change ZAP mode. Mode can now be changed even if the main toolbar is hidden.
- Used checkboxes for mode selection instead of a radio box to maintain GUI consistency.
- There's one issue, however. The display in the MainToolBar panel which shows the ZAP mode and allows for user to select from a drop down list doesn't change when the user changes the mode from the Edit menu. I couldn't to do this as I wasn't able to figure out how to get a reference to the [toolbar](https://github.com/zaproxy/zaproxy/blob/96e98a3a73b2ba7fcfc86eb78cf6518937f7eb40/src/org/zaproxy/zap/view/MainToolbarPanel.java) from the [ExtensionState.java](https://github.com/zaproxy/zaproxy/blob/0879ae7de05dcb04b50f71ce4b7551e8b29977cf/src/org/parosproxy/paros/extension/state/ExtensionState.java) class. If anyone could help me out  with that, I am willing to finish that as well.

Please review the PR and suggest changes if any. Thanks.

kingthorin commented on src/org/parosproxy/paros/extension/state/ExtensionState.java:
```diff
@@ -67,11 +75,46 @@ public void hook(ExtensionHook extensionHook) {
+			}
+			modeCount = 0;
+			for(Mode modeType : Mode.values()){
+			addZAPModeEventListeners(modeType,menuItemsMode[modeCount],menuItemsMode);
```
Could this not be done inside the loop above? Instead of looping through the same elements twice?

rajiv65 replied:
I need all elements of my array of checkboxes i.e. menuItemsMode[] to be set before I pass the array as an argument to my addZAPModeEventListeners() function. This is because whenever I click on either of the modes, I am first deselecting all the checkboxes and then checking the correct one. So, I need all my menu items to be there before I add any eventListener and hence, before adding event listeners, I am first adding all the checkbox items and then I am looping again to add eventListeners. It can be done easily with a radio button type of menu but then the GUI would look different compared to other menu items in ZAP.

kingthorin replied:
Fair enough.

thc202 replied:
It could be used a `ButtonGroup` instead, which already takes care to keep just one of the buttons selected.

kingthorin commented on src/org/parosproxy/paros/extension/state/ExtensionState.java:
```diff
@@ -67,11 +75,46 @@ public void hook(ExtensionHook extensionHook) {
+
+	private void addZAPModeEventListeners(final Mode modeType, final JCheckBoxMenuItem checkBox, final JCheckBoxMenuItem[] menuItemsMode){
+		 checkBox.addActionListener(new java.awt.event.ActionListener() {
+		 public void actionPerformed(java.awt.event.ActionEvent evt) {
```
Should have an `@Override` annotation.

rajiv65 replied:
Sorry but I couldn't get you. As far as I know, @Override is used when we override a parent class method. I don't think I have overriden any other parent method here. Correct me if I am wrong. Thanks.

kingthorin replied:
I don't have an IDE in front of me.  But, aren't you overriding actionPerformed on the checkbox? I'm kind of trusting google here (since no IDE).... http://www.codejava.net/java-se/swing/jcheckbox-basic-tutorial-and-examples#AddListeners

kingthorin commented:
Please ensure you change is documented with a //ZAP comment since this modifies a Paros source file (ExtensionState.java).
Per https://github.com/zaproxy/zaproxy/wiki/DevGuidelines .

As for updating the MainToolBar....I'm guessing: Check `View.isInitialised()`, then `View.getSingleton().getMainFrame()`, get the ToolbarPanel (https://github.com/zaproxy/zaproxy/blob/96e98a3a73b2ba7fcfc86eb78cf6518937f7eb40/src/org/parosproxy/paros/view/MainFrame.java#L308) get the toolbar and setMode on it? (https://github.com/zaproxy/zaproxy/blob/96e98a3a73b2ba7fcfc86eb78cf6518937f7eb40/src/org/zaproxy/zap/view/MainToolbarPanel.java#L187)

rajiv65 commented:
I will look into the toolbar as you have suggested and also document the change of the Paros source file and get back to you if there's any issue. Thanks.

kingthorin commented on src/org/parosproxy/paros/extension/state/ExtensionState.java:
```diff
@@ -67,11 +75,46 @@ public void hook(ExtensionHook extensionHook) {
+			int modeCount = 0;
+			for (Mode modeType : Mode.values()) {
+			menuItemsMode[modeCount] = addZAPModeMenuItem(modeType);
+  		menuZAPMode.add(menuItemsMode[modeCount]);
```
Something going on with the indentation here...

rajiv65 replied:
Didn't look like that in my local copy. Not sure how that happened. Guess my text editor is at fault here. I'll change this.

kingthorin replied:
Could be spaces vs tabs within the file, depending on your editor's setting.

rajiv65 commented:
Please review the latest changes. Thanks.
- Added the override annotation for the actionPerformed() method in the event listener.
- Documented the change in the ExtensionState.java file as per the ZAP guidelines for a change in Paros source files.
- The ZAP mode change from edit menu is reflected in the toolbar as well now.
- Fixed indentation issues.

rajiv65 commented:
@kingthorin Anything else you want me to add or change?

kingthorin commented:
One minor thing, the indentation of the loops, the loops should actually "contain" the code....know what I mean?

Then would ask that you do an interactive rebase and fix-up or Squash into a single "atomic" commit: https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#what-we-zap-team-expect-from-you .

- https://www.kernel.org/pub/software/scm/git/docs/git-rebase.html#_interactive_mode
- https://wiki.eclipse.org/EGit/User_Guide#Interactive_Rebase

@psiinon, @thc202 do you see anything that needs to be changed/addressed?

rajiv65 commented:
Yes. I got what you said. Will indent the code inside the loops. And sorry about all these things that I am overlooking. I am relatively new to open source contribution and am still learning. Thanks for pointing out my mistakes. :)

thc202 commented on src/org/parosproxy/paros/extension/state/ExtensionState.java:
```diff
@@ -67,11 +77,50 @@ public void hook(ExtensionHook extensionHook) {
+	}
+
+	private JCheckBoxMenuItem addZAPModeMenuItem(Mode modeType){
+		final JCheckBoxMenuItem modeItem = new JCheckBoxMenuItem(modeType.name().substring(0,1).toUpperCase()+modeType.name().substring(1)+" Mode");
```
This should be internationalised, since there are already resource messages (e.g. "view.toolbar.mode.safe.select") it could use those.

thc202 commented:
I don't see any problem using `JRadioButtonMenuItem`, that's the expected UI component when only one of the options can be selected at any given time.

The selected state of the modes in the edit menu does not reflect the correct mode when changed through the main tool bar.

Why `ExtensionState`? That extension is about/for session (cookie) management, which is unrelated to the `Mode`, moreover if the extension is disabled the option would not be shown. Since the `Mode` is a built-in core functionality it should probably be added by the `View` (or other "core" class).

thc202 commented:
Also, the class CoreAPI might need some changes (when processing the endpoint `ACTION_SET_MODE`).
The `ExtensionUiUtils` could be changed to update the state of the UI components when the `Mode` changes (instead of having some of the code conditionally calling the UI components directly).

kingthorin commented:
@rajiv65 no need to apologize at all. Lots of details and it always takes time to get used to.

As you can see form @thc202's comments there are things I missed too :smile:

thc202 commented on src/org/parosproxy/paros/extension/state/ExtensionState.java:
```diff
@@ -52,6 +60,8 @@
 	private JCheckBoxMenuItem menuSessionTrackingEnable = null;
 
 	private ZapMenuItem menuResetSessionState = null;
+	private JCheckBoxMenuItem[] menuItemsMode = new JCheckBoxMenuItem[Mode.values().length];
```
UI related components should not be initialised unless ZAP has GUI (there's the `initView` method for these kind of initialisations or in the `hook` method, if there's a view).

thc202 commented on src/org/parosproxy/paros/extension/state/ExtensionState.java:
```diff
@@ -32,13 +32,21 @@
 import javax.swing.JCheckBoxMenuItem;
 import javax.swing.JOptionPane;
+import javax.swing.JMenu;
+import javax.swing.event.MenuEvent;
```
Unused imports should be removed or not introduced (`MenuEvent`, `MenuListener` and the added `Control.Mode`).

rajiv65 commented:
Okay. So I have got rid of the redundant checkboxes and used radio buttons and button group as @thc202 suggested. I also incorporated the other changes as suggested. Just one more thing.
Will [ExtensionEdit.java](https://github.com/zaproxy/zaproxy/blob/0879ae7de05dcb04b50f71ce4b7551e8b29977cf/src/org/parosproxy/paros/extension/edit/ExtensionEdit.java) be a good place to add this change ZAP mode menu item? I see that the **Find option** of the Edit menu is added here, which I believe is also a built-in core functionality just like ZAP mode. If not, where do you suggest I add the code for this feature.

thc202 commented:
I'd add that in `MainMenuBar` (as the combo box is directly added to/by the `MainToolbarPanel`), that prevents removing/hiding the mode menu by disabling an extension.

rajiv65 commented:
Please review the latest commit @thc202  @kingthorin . I'll squash all the commits into one once I incorporate any other suggested changes given by you, if any. Thanks.

- Shifted code to add "change ZAP mode" option of edit menu to [MainMenuBar.java](https://github.com/zaproxy/zaproxy/blob/296801bb838ae1ceca102a6be5b5ed2e8c29e097/src/org/parosproxy/paros/view/MainMenuBar.java) 
- Used radio buttons and button group instead of checkboxes
- Selected state of the modes in the edit menu now reflects the correct mode when changed through the main tool bar.
- Edited CoreAPI class for processing the endpoint ACTION_SET_MODE
- Internationalised menu items names using resource messages
- Removed unnecessary imports
- Documented the change in the MainMenuBar.java file as per the ZAP guidelines for a change in Paros source files.

kingthorin commented on src/org/parosproxy/paros/extension/state/ExtensionState.java:
```diff
@@ -52,7 +52,7 @@
 
 	private ZapMenuItem menuResetSessionState = null;
-
+	
```
If you could revert these formatting changes that'd be great.

https://github.com/zaproxy/zaproxy/wiki/DevGuidelines

kingthorin commented on src/org/parosproxy/paros/extension/state/ExtensionState.java:
```diff
@@ -71,7 +71,6 @@ public void hook(ExtensionHook extensionHook) {
 		    extensionHook.getHookMenu().addEditMenuItem(getMenuResetSessionState());
 	    }
 	}
-
```
Same

rajiv65 commented:
Reverted the formatting changes. Please review the remaining code and let me know if any changes are to be done. @thc202 @kingthorin

kingthorin commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -75,6 +87,8 @@
 	private ZapMenuItem menuHelpAbout = null;
     private JMenu menuAnalyse = null;
+    private JMenu menuZAPMode = null;
+    ButtonGroup menuZAPModeGroup = null;
```
private ?

rajiv65 replied:
Forgot to add. Did it now.

kingthorin commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -114,10 +128,59 @@ private void initialize() {
+	}
+
+	private JRadioButtonMenuItem addZAPModeMenuItem(final Mode modeType){
+		final JRadioButtonMenuItem modeItem = new JRadioButtonMenuItem(""+Constant.messages.getString("view.toolbar.mode."+modeType.name()+".select"));
```
Is the leading empty string necessary in the concatenation?

rajiv65 replied:
Not necessary. Removed it now.

kingthorin commented on src/org/zaproxy/zap/view/MainToolbarPanel.java:
```diff
@@ -178,6 +178,9 @@ public void actionPerformed(java.awt.event.ActionEvent e) {
 					}
 					Control.getSingleton().setMode(mode);
+					if (View.isInitialised()){
+	 		    		View.getSingleton().getMainFrame().getMainMenuBar().setMode(mode);
```
This should be indented

rajiv65 replied:
Done.

kingthorin commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -114,10 +128,59 @@ private void initialize() {
+	}
+
+
+
```
Two of these lines could be removed

rajiv65 replied:
Removed the unnecessary lines.

rajiv65 commented:
Added the necessary changes. Anything else that needs to be addressed? Thanks.

kingthorin commented:
Just committing those changes and fixing this up into a single commit.

https://github.com/zaproxy/zaproxy/pull/2837#issuecomment-243742034

rajiv65 commented:
@thc202  @kingthorin  Squashed all changes into one commit. Please review it and let me know. Thanks.

kingthorin commented:
Looks like you missed one. It is down from 4 to 2, but it's still 2 :smiley: 

Code wise it looks ok to me.

rajiv65 commented:
One of the commit is the merge of my forked repository with upstream. That's why I left it as it is. Do you want me squash that one too?

thc202 commented:
Yes, please, prefer rebase to merge. The commit message could be tweaked a little (e.g. some of the phrases could be removed).

I'll add some review comments shortly.

rajiv65 commented:
Squashed the 2 commits into one now.

rajiv65 commented:
@thc202  @kingthorin  Are there any other changes I should be making? Do let me know. Thanks.

thc202 commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -41,7 +44,16 @@
 import javax.swing.JOptionPane;
 import javax.swing.KeyStroke;
-
+import javax.swing.JMenu;
```
This unused import could be removed.

thc202 commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -41,7 +44,16 @@
+
+import org.parosproxy.paros.control.Control.Mode;
+import org.parosproxy.paros.view.View;
+import org.parosproxy.paros.control.Control;
```
This class was already imported.

thc202 commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -114,10 +128,57 @@ private void initialize() {
 			menuEdit = new javax.swing.JMenu();
 			menuEdit.setText(Constant.messages.getString("menu.edit")); // ZAP: i18n
 			menuEdit.setMnemonic(Constant.messages.getChar("menu.edit.mnemonic"));
+			menuEdit.add(getMenuEditZAPMode());
```
Might worth add a separator after this entry.

thc202 commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -114,10 +128,57 @@ private void initialize() {
+		if (menuZAPMode == null) {
+			menuZAPMode = new JMenu(Constant.messages.getString("menu.edit.zapmode"));
+			menuZAPModeGroup = new ButtonGroup();
+			JRadioButtonMenuItem newButton = null;
```
The null initialisation can be removed.

thc202 commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -114,10 +128,57 @@ private void initialize() {
+			menuZAPModeGroup = new ButtonGroup();
+			JRadioButtonMenuItem newButton = null;
+			for (Mode modeType : Mode.values()) {
+				 newButton = addZAPModeMenuItem(modeType);
```
In this and following two lines there are some spaces at the beginning of the statements and between the tabs.

thc202 commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -114,10 +128,57 @@ private void initialize() {
+		final JRadioButtonMenuItem modeItem = new JRadioButtonMenuItem(Constant.messages.getString("view.toolbar.mode."+modeType.name()+".select"));
+		modeItem.addActionListener(new java.awt.event.ActionListener() {
+		@Override
+		public void actionPerformed(java.awt.event.ActionEvent evt) {
```
This method should be indented one more level to the right.

thc202 commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -114,10 +128,57 @@ private void initialize() {
+		modeItem.addActionListener(new java.awt.event.ActionListener() {
+		@Override
+		public void actionPerformed(java.awt.event.ActionEvent evt) {
+	 		Control.getSingleton().setMode(modeType);
```
Same comment regarding the spaces between the tabs.

thc202 commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -114,10 +128,57 @@ private void initialize() {
+		@Override
+		public void actionPerformed(java.awt.event.ActionEvent evt) {
+	 		Control.getSingleton().setMode(modeType);
+	 		if (View.isInitialised()){
```
The if statement can be removed, the view is initialised (it's view code that will call this method).

thc202 commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -114,10 +128,57 @@ private void initialize() {
+			@Override
+			public void run() {
+				Enumeration<AbstractButton> elements = menuZAPModeGroup.getElements();
+				AbstractButton button = null;
```
The null initialisation can be removed.

thc202 commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -114,10 +128,57 @@ private void initialize() {
+				Enumeration<AbstractButton> elements = menuZAPModeGroup.getElements();
+				AbstractButton button = null;
+				while (elements.hasMoreElements()) {
+      				   button = elements.nextElement();
```
Mixed indentation chars (tabs and spaces).

thc202 commented on src/org/parosproxy/paros/view/MainMenuBar.java:
```diff
@@ -114,10 +128,57 @@ private void initialize() {
+				AbstractButton button = null;
+				while (elements.hasMoreElements()) {
+      				   button = elements.nextElement();
+      				   if (button.getText().toLowerCase().contains(mode.name().toString().toLowerCase())) {
```
This will not work properly when the modes are translated (already are in some of the languages).

rajiv65 replied:
Any suggestions on how do I go about correcting this?

thc202 replied:
It could use a `Map<Mode, JRadioButtonMenuItem>`, then it just needs to `get` the menu (instead of iterate the menus).

thc202 commented on src/org/zaproxy/zap/extension/api/CoreAPI.java:
```diff
@@ -429,6 +429,7 @@ public void run() {
 				Mode mode = Mode.valueOf(params.getString(PARAM_MODE).toLowerCase());
 		    	if (View.isInitialised()) {
 	    			View.getSingleton().getMainFrame().getMainToolbarPanel().setMode(mode);
+	    			View.getSingleton().getMainFrame().getMainMenuBar().setMode(mode);
```
This has side effect of setting the mode twice (same when changing the mode through the Edit menu). The method Control.setMode(Mode) could be changed to ignore the mode (e.g. not notify of the change nor persist...) when already set.

thc202 commented on src/org/zaproxy/zap/view/MainToolbarPanel.java:
```diff
@@ -178,6 +178,9 @@ public void actionPerformed(java.awt.event.ActionEvent e) {
 						default: return;	// Not recognised
 					}
 					Control.getSingleton().setMode(mode);
+					if (View.isInitialised()){
```
The if statement can be removed, the view is initialised (it's view code that will call this method).

thc202 commented:
Comments added, sorry for the delay.

thc202 commented:
btw, the commit message still has redundant/unnecessary info (e.g. one close is enough, some info is not interesting for the history).

kingthorin commented:
@rajiv65 are you able to address @thc202's feedback?

kingthorin commented:
Could you set this PR (branch) to allow modification by maintainers?
https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

kingthorin commented:
@rajiv65 if you're unable to address the feedback could you please set this PR (branch) to allow modification by maintainers?
https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

kingthorin commented:
I guess this is abandon?

kingthorin commented:
Tried to address conflict on this, couldn't push the changes back to the user repo. Have sent an email.

thc202 commented:
@kingthorin open a new pull request with the latest changes.
